src/metabase/models/dashboard_tab.clj

```diff
@@ -59,19 +61,100 @@
+
+;;; -------------------------------------------------- CRUD fns ------------------------------------------------------
+
+(mu/defn create-tabs! :- [:map-of neg-int? pos-int?]
+  "Create the new tabs and returned a mapping from temporary tab ID to the new tab ID."
+  [dashboard-id :- ms/PositiveInt
+   new-tabs     :- [:sequential [:map [:id neg-int?]]]]
+  (let [new-tab-ids (t2/insert-returning-pks! :model/DashboardTab (->> new-tabs
+                                                                       (map #(dissoc % :id))
+                                                                       (map #(assoc % :dashboard_id dashboard-id))))]
+    (zipmap (map :id new-tabs) new-tab-ids)))
+
+(mu/defn update-tabs! :- nil?
+  "Updates tabs of a dashboard if changed."
+  [current-tabs :- [:sequential [:map [:id ms/PositiveInt]]]
+   new-tabs     :- [:sequential [:map [:id ms/PositiveInt]]]]
+  (let [update-ks       [:name :position]
+        id->current-tab (m/index-by :id current-tabs)
+        to-update-tabs  (filter
+                          ;; filter out tabs that haven't changed
+                          (fn [new-tab]
+                            (let [current-tab (get id->current-tab (:id new-tab))]
+                              (not= (select-keys current-tab update-ks)
+                                    (select-keys new-tab update-ks))))
+
+                          new-tabs)]
+    (doseq [tab to-update-tabs]
+      (t2/update! :model/DashboardTab (:id tab) (select-keys tab update-ks)))
+    nil))
+
+(mu/defn delete-tabs! :- nil?
+  "Delete tabs of a Dashboard"
+  [tab-ids :- [:sequential {:min 1} ms/PositiveInt]]
+  (when (seq tab-ids)
+    (t2/delete! :model/DashboardTab :id [:in tab-ids]))
+  nil)
+
+(mu/defn classify-changes :- [:map
+                              [:to-create [:maybe [:sequential [:map [:id ms/Int]]]]]
+                              [:to-update [:maybe [:sequential [:map [:id ms/Int]]]]]
+                              [:to-delete [:maybe [:sequential [:map [:id ms/Int]]]]]]
+  "Given 2 lists of seq maps of changes, where each map an has an `id` key,
+  return a map of 3 keys: `:to-create`, `:to-update`, `:to-delete`.
+
+  Where:
+    :to-create is a list of maps that ids in `new-items`
+    :to-update is a list of maps that has ids in both `current-items` and `new-items`
+    :to delete is a list of maps that has ids only in `current-items`"
+  [current-items :- [:maybe [:sequential [:map [:id ms/PositiveInt]]]]
+   new-items     :- [:maybe [:sequential [:map [:id ms/Int]]]]]
+  (let [[delete-ids create-ids update-ids] (diff (set (map :id current-items))
+                                                 (set (map :id new-items)))]
+    {:to-create (when (seq create-ids) (filter #(create-ids (:id %)) new-items))
+     :to-delete (when (seq delete-ids) (filter #(delete-ids (:id %)) current-items))
+     :to-update (when (seq update-ids) (filter #(update-ids (:id %)) new-items))}))
+
+(defn do-update-tabs!
```

Comment:
Found a bug. Imagine you start with no tabs, then add some cards to the dash. Save. This is version A. Then you create a tab, save, then delete it, save. So now you have one tab with some cards. Then revert to version A. The one remaining tab should get deleted, but the dashcard’s dashboard_tab_id ID field won’t update, and will continue referencing the deleted tab.

Reply:
I'm not sure I'm following, can you loom it?

but I don't think it's possible for dashcard to reference to deleted tabs since we have an FK for it.

Reply:
Ah of course, you're right. Then the weird thing is the dashcards on the first tab will be deleted and will have to be created again, along with any associated series they have. This is probably OK as long as the dashcards are updated after the tabs are deleted. Not sure if it's worth complicating the code to fix this though, because it works and can't see how it could break.